### DISCUSSION TITLE
How can I model an x-ray system in EGSnrc?

### DISCUSSION THREAD

**Discussion by bdnkth** [nrc-cnrc/EGSnrc#797] [Questions] [answered]
Hi,

I'm pretty new to the world of EGSnrc. For our upcoming xray project we want to run simulations in order to design our radiation protection.

I've started implementing our machine in the egs world which so far has worked pretty well. However, when it comes to simulating the radiation, the calculated values seem way to low. I've tried to fix this but so far I was not able to. So hopefully someone here can help me.

I'm currently using an implementation which is based on the egs_chamber example. I've stumpled across the BEAMnrc package which seems to be better for xray but is not as user friendly as the egs_chamber stuff. Can I do my simulations with the egs_chamber or do I have to switch to BEAMnrc? Our xray source can handle up to 70 kV so that's the maximum which I have to simulate. Currently I'm using our measured spectrum of the xray source as the radiation source in egsnrc.

**Comment by rtownson** [comment]
Either egs_chamber or BEAMnrc can be used for this, and both have some advantages or disadvantages. Since you're already working with egs_chamber, you might as well stay there. They use the same transport physics and data! 

Can you provide some numbers - how do the experimental numbers compare? Note that MC simulations are best used for relative comparisons, and all results are reported "per source fluence" or "per initial particle". Calculating absolute dose with simulations is generally not as reliable, since in the real world you don't know exactly how many source particles there were.

**Comment by bdnkth** [comment]
Good to hear that egs_chamber also works.

```
1 PB700ICRU       11.340   3430.0000 3.0951e-03 +/- 0.240  % 1.2747e-17 +/- 0.240  %
```

These are the results for a lead plate I've placed approximately 5cm below the radiation source. For our understanding the dose area product is really small.

**Reply by rtownson** [reply]
What are you comparing to? Like I said, this is "dose per source particle", so it will always be a very small number.

**Comment by ftessier** [comment]
@bdnkth, the expected dose result would be on the order of the energy deposited in the lead by a _single_ representative source photon, divided by the mass of the lead plate (in Joule/kg). Here your results show that the energy deposited on average is `3.0951e-03 +/- 0.240  %` MeV, so about 3 keV. Does this make sense, theoretically, according to the cross section of ~70 keV photons incident on a thin lead plate?

**Reply by ftessier** [reply]
What is the thickness of your lead plate?

**Reply by ftessier** [reply]
Presumably, the average energy of the photons is also _much_ less than 70 keV, if that is the maximum x-ray spectrum energy. What does the spectrum look like?

**Reply by ftessier** [reply]
Also, if you are using an isotropic source, the results are normalized by the total number of emitted photons, _in `4*pi`._ To get a result normalized by fluence within a given aperture, you can use a collimated source.

**Reply by ftessier** [reply]
You may also want to look at particle tracks in `egs_view` to get a sense, visually, of whether or not the lead shield is stopping photons, and to what extent.

**Comment by bdnkth** [comment]
The lead plate is defined as box size = 64.7 45.7 1 and the source is above the plate in z-dimension. That the deposited energy is from a single source photon explains why I was expecting a higher dosis.

I'm currently using a pencil_beam source, I'll try with something different.

I've attached the used [spectrum](https://github.com/nrc-cnrc/EGSnrc/files/7569752/tungsten70.txt)

**Reply by ftessier** [reply]
Sounds good. A pencil beam is fine, I was just worried that you were using an isotropic source, which would decreased the dose per photon on account of the particles that are emitted away from the plate.

Results are normalized by particle (or fluence), because we want the results to converge. We typically adjust the number of histories to reach the sought level of uncertainty, and we don't want the dose value to change all the time, depending on _N._ The history number is in fact the only unit of "time" in these simulations, so you can view the results as rates (per particle emitted).

**Comment by bdnkth** [comment]
Okay, that makes sense. Since the deposited energy is per photon, is there a way to calculate the number of particles hitting the plate? I guess there is, probably I can't see the forest for the trees

**Comment by bdnkth** [comment]
Any updates on this? I'm still struggling to find out the number of photons hitting my object

**Reply by rtownson** [reply]
You might want to try scoring a phase-space on the surface where particles enter the plate. Then you can use beamdp_gui to analyze the phase-space (e.g. calculate fluence), or just look at the number of photons in it, if that's really what you want.

https://nrc-cnrc.github.io/EGSnrc/doc/pirs898/classEGS__PhspScoring.html

**Reply by bdnkth** [reply]
In the end I need to simulate how much energy/radiation is in a certain part of my model. So I assume I have to so sum up the energy of all photons, right?

**Reply by ftessier** [reply]
Yes, but normally you would score dose in the region of interest. In `egs_chamber`, you specify a cavity geometry where dose is scored. You can set the cavity mass to `1000` grams (1 kg), then you will get the energy deposited (in Joules). Alternatively, you can use a dose scoring object and get the dose in all regions.

**Reply by ftessier** [reply]
Beyond that you would modify the `egs_chamber.cpp` application code to suit your specific need.

**Reply by bdnkth** [reply]
Cavity geometry looks like the thing I was searching for. Spend the better part of today reading the documentation for it and trying to implement it in my code. I'm not really sure if I can use the cavity geometry scoring in egs_chambers or if I have to use it in the cavity code. It compiles and runs at least in my egs_chamber code, however, the output looks like this:
Running 100000000 histories

  Batch             CPU time        Result   Uncertainty(%)
==========================================================
      1                6.16              0         100.00
      2               12.28              0         100.00
      3               18.28              0         100.00
      4               24.51              0         100.00
      5               30.61              0         100.00
      6               36.60              0         100.00
      7               42.59              0         100.00
      8               48.81              0         100.00
      9               55.03              0         100.00
     10               61.09              0         100.00

For the sake of sanity I tried on the example inputs in the cavity folder, however, they run into some libBeam missing file erro

**Reply by ftessier** [reply]
Unless you need specific features of `cavity`, use `egs_chamber`, which is more recent and has features of its own for scoring dose in cavity regions (see the [egs_chamber paper](https://github.com/nrc-cnrc/EGSnrc/wiki/Citing-EGSnrc#egs_chamber) for more information).

The results you are showing suggest that no particle is depositing dose in the regions of interest, or else the region is so small (thin?) and the density so low (air?) that no photon interacts there in the course of the simulation.

The issue with the example1 file is noted, indeed it relies on a BEAMnrc simulation source or materials which are not part of the distribution; I added this as issue #804

**Reply by bdnkth** [reply]
I've tried to figure out why theres always a zero result. Unfortunatly, I have not been successfully. I've attached a simplified version of the model (the main part is the same), so you might have a look what's going wrong
[test.txt](https://github.com/nrc-cnrc/EGSnrc/files/7649925/test.txt)
.

**Reply by ftessier** [reply]
The geometry lies on the `+z` axis, but the particles are emitted along `-z`: `direction = 0 0 -1`.

**Reply by ftessier** [reply]
Oh I see you are emitting towards the thin steel plate, which is in `-z`. The issue is that you have to set the calculation geometry to `machine`, instead of `below_source`:

```
    :start calculation geometry:
        geometry name = machine
        cavity regions = 8
        cavity mass = 100000
    :stop calculation geometry:
```

The `geometry name` is the name of the entire composite geometry, and then you select the scoring regions (the `cavity regions`) within that geometry.

**Reply by ftessier** [reply]
Note that you cannot have strictly overlapping planes in egs++, from two disjoint geometries. These will show up as speckle pattern, showing when particles hit or miss the second plane (owing to the limit of floating-point precision). For example, in you geometry:

![image](https://user-images.githubusercontent.com/1444241/144614359-b661a2bf-1681-4729-b41e-80d0d219de2c.png)

**Reply by ftessier** [reply]
Another note: the `union` is not a very efficient geometry, because at every step the code must calculate the distance to **all** geometries in the union, since in general there is not constraint on where they are located with respect to each other.

**Reply by ftessier** [reply]
Here is a more efficient way to build the main box:

```
    :start geometry:
        name = conveyor_channel
        library = egs_ndgeometry
        type = EGS_XYZGeometry
        x-planes = -35.15 -34.95 34.95 35.15
        y-planes = -64 -63.7 63.7 64
        z-planes = 0 0.25 0.75 2 29.7 30
        :start media input:
            media = STEEL700ICRU AIR700ICRU
            set medium = 4 1
            set medium = 22 1
            set medium = 28 1
            set medium = 31 1
            set medium = 34 1
        :stop media input:
    :stop geometry:
    simulation geometry = conveyor_channel
```

![image](https://user-images.githubusercontent.com/1444241/144619046-972f4ebe-306f-47c6-91c8-2a0b703ff0d6.png)

(clipped to show internal structure)

**Comment by ftessier** [comment, marked as answer]
@bdnkth here is a better way to contruct your geometry, without unions. The idea is to build the shield, the conveyor, and the radiation box, and then "glue" them together as three layers along z using a CD geometry (there is a sligh complication because the radiation box has to "punch" through the conveyor):

```
:start geometry definition:

    ## big air box
    :start geometry:
        name     = air_box
        library  = egs_box
        box size = 200  200 400
        :start media input:
            media = AIR700ICRU
        :stop media input:
    :stop geometry:

    ############################################################################v
    ### conveyor
    ############################################################################

    ### conveyor channel
    :start geometry:
        name     = conveyor
        library  = egs_ndgeometry
        type     = EGS_XYZGeometry
        x-planes = -35.15 -34.95 34.95 35.15
        y-planes = -64 -63.7 63.7 64
        z-planes = -40 0.25 0.75 2 29.7 40
        :start media input:
            media = STEEL700ICRU AIR700ICRU
            set medium = 4 1
            set medium = 22 1
            set medium = 28 1
            set medium = 31 1
            set medium = 34 1
        :stop media input:
    :stop geometry:

    ### hole in top
    :start geometry:
        name     = hole
        library  = egs_ndgeometry
        type     = EGS_XYZGeometry
        x-planes = -32.2 32.2
        y-planes = -22.7 22.7
        z-planes = 29 39
        :start media input:
            media = AIR700ICRU
        :stop media input:
    :stop geometry:

    ### conveyor with hole
    :start geometry:
        name    = hole_in_conveyor
        library = egs_genvelope
        base geometry = conveyor
        inscribed geometries = hole
    :stop geometry:

    ### conveyor in air
    :start geometry:
        name    = converyor_box_in_air
        library = egs_genvelope
        base geometry = air_box
        inscribed geometries = hole_in_conveyor
    :stop geometry:


    ############################################################################
    ### radiation box
    ############################################################################

    ### radiation box
    :start geometry:
        name     = radiation_box
        library  = egs_ndgeometry
        type     = EGS_XYZGeometry
        x-planes = -32.5 -32.2 32.2 32.5
        y-planes = -23 -22.7 22.7 23
        z-planes = 0 85.2 85.5
        :start media input:
            media = STEEL700ICRU AIR700ICRU
            set medium = 4 1
        :stop media input:
    :stop geometry:

    ### radiation box in air
    :start geometry:
        name    = radiation_box_in_air
        library = egs_genvelope
        base geometry = air_box
        inscribed geometries = radiation_box
    :stop geometry:


    ############################################################################
    ### shield
    ############################################################################

    ### shield
    :start geometry:
        name     = shield
        library  = egs_ndgeometry
        type     = EGS_XYZGeometry
        x-planes = -35.15 35.15
        y-planes = -15 15
        z-planes = -1 10
        :start media input:
            media = PB700ICRU
        :stop media input:
    :stop geometry:

    ### shield in air
    :start geometry:
        name    = shield_in_air
        library = egs_genvelope
        base geometry = air_box
        inscribed geometries = shield
    :stop geometry:


    ############################################################################
    ### machine
    ############################################################################

    ### machine planes
    :start geometry:
        name      = machine_planes
        library   = egs_planes
        type      = EGS_Zplanes
        positions = -100 0 30 100
    :stop geometry:

    ### machine
    :start geometry:
        name    = machine
        library = egs_cdgeometry
        base geometry = machine_planes
        set geometry = 0 shield_in_air
        set geometry = 1 converyor_box_in_air
        set geometry = 2 radiation_box_in_air
    :stop geometry:

    # ### simulation geometry
    simulation geometry = machine

:stop geometry definition:
```

Here is a snapshot (with energy set to 300 keV to get some tracks, and cavity regions set to `1`):

![image](https://user-images.githubusercontent.com/1444241/144632674-20aaccb6-f34a-4ce7-9521-78a0fa1db2d1.png)

**Reply by ftessier** [reply]
```
Running 1000000 histories

  Batch             CPU time        Result   Uncertainty(%)
==========================================================
      1                0.56     1.3614e-14           0.07
      2                1.11    1.36192e-14           0.05
      3                1.66    1.36209e-14           0.04
      4                2.21    1.36188e-14           0.04
      5                2.76    1.36255e-14           0.03
      6                3.30    1.36255e-14           0.03
      7                3.85    1.36273e-14           0.03
      8                4.40    1.36265e-14           0.03
      9                4.95    1.36259e-14           0.02
     10                5.49     1.3626e-14           0.02


Finished simulation

Total cpu time for this run:            5.50 (sec.) 0.0015(hours)
Histories per hour:                     6.54545e+08   
Number of random numbers used:          150854217     
Number of electron CH steps:            5.50715e+06   
Number of all electron steps:           5.52284e+06   


 last case = 1000000 fluence = 33333.3

Geometry                        Cavity dose      
-----------------------------------------------
machine                   1.3626e-14 +/- 0.023  % 

```

**Comment by bdnkth** [comment]
Tank you for that lenthly response. I think I'll need some time to process all that information. Originally I've started with the xyzgeometries but I've struggled with their syntax and switched to the boxes since they are really straight forward to work with.

Can you check if this [file](https://github.com/nrc-cnrc/EGSnrc/files/7661413/xyzaproach.txt) connected all your code in the right way? Since I do not get the same result as you

**Reply by ftessier** [reply]
I don't think there is a `position` input for the box shape (use a translation `transformation` instead). I am not sure what is the intent with the source as defined. Can you share an `egs_view` screenshot with a few tracks? Our results would be different because I used a narrow beam starting at the origin and aimed along `-z`, just to show some dose was deposited in the lead plate.

**Reply by ftessier** [reply]
Agreed, the `xyz` geometry is not immediately obvious for this kind of setup; but it is the most efficient for anything that can be drawned on a rectilinear grid. It also makes it easier to change, say, the thickness of a wall by changing a single number, instead of changing the wall thickness, and then adjusting the position of the boxes. The fundamental issue with boxes is that the faces of two different boxes cannot overlap on the same plane. The alternative, to build a box shell, is to inscribe a box of air inside a box of the wall material, but then it is not obvious how to punch holes through the faces (as was done in my example, to fuse the radiation box on top of the conveyor).

**Reply by bdnkth** [reply]
I would love to show you some tracks from the .ptracks file but it is not created when I run the simulation. Is this because I have all zeros in my result? I changed from position to transformation but that did not change anything.

**Reply by bdnkth** [reply]
Still getting zeros as result without any clue. I've removed the transformation for the source to get closer to your simulation but it did not change the result.

**Reply by bdnkth** [reply]
@ftessier any clues?

**Reply by rtownson** [reply]
When you changed the position input to a transformation did you include the entire transformation input block in your shape input?
https://nrc-cnrc.github.io/EGSnrc/doc/pirs898/classEGS__AffineTransform.html

e.g.

```
        :start shape:
            type = box
            box size = 3 3 1
            :start transformation:
                translation =  0 -21.125 41.5 # in cm
            :stop transformation:
        :stop shape:
```

**Reply by bdnkth** [reply]
Yes I did. I also tried it without having the transformation, to see if that's the problem, but somehow the problem prevailed. I've attached the file, to show it
[currentStatus.txt](https://github.com/nrc-cnrc/EGSnrc/files/7728262/currentStatus.txt)
.

**Reply by rtownson** [reply]
Are you running the simulation using `egs_app`? The inputs you have for scoring, under `:start scoring options:` are only used by `egs_chamber`. Your input file gives results for me when I use `egs_chamber`.

**Reply by bdnkth** [reply]
I guess that's the problem. I was laking some code from the egs_chamber.cpp in my project cpp file which I now added. But then make failes with
Mortran compiling EGSnrc sources ...
gfortran -O2 -mtune=native -fPIC    -c -o egsnrc_linux64.o egsnrc_linux64.F
egsnrc_linux64.F:8:9:

    8 |       d %C80
      |         1
Error: Symbol ‘d’ at (1) has no IMPLICIT type
make: *** [/mnt/c/Users/Kath/EGSnrc/HEN_HOUSE/makefiles/cpp_makefile:70: egsnrc_linux64.o] Error 1

**Reply by rtownson** [reply]
I'm not sure why you're seeing this, there must be a typo or bug in the code you have added. Why are you writing your own code, what are you trying to do that egs_chamber can't?
